falcon/routing/static.py

```diff
@@ -49,10 +55,15 @@ def __init__(self, prefix, directory, downloadable=False):
         self._prefix = prefix
         self._directory = directory
         self._downloadable = downloadable
+        if default_filename is not None:
+            self._default_filename = os.path.join(directory, default_filename)
+        else:
+            self._default_filename = None
 
     def match(self, path):
         """Check whether the given path matches this route."""
-        return path.startswith(self._prefix)
+        prefix = self._prefix if self._default_filename is None else self._prefix[:-1]
```

Comment:
I don't think this quite addresses the issue. Consider the case in which `self._default_filename is not None` and:

```py
self._prefix = '/foo/bar/'
path = '/foo/barley'
```
In this case, `prefix = '/foo/bar'`, which would result in `match()` returning `True`.

(P.S. Please add a test for this scenario.)